design/erasure-coding.md

```diff
@@ -0,0 +1,83 @@
+
+
+The algorithm looks like the following:
+
+1) Given a dataset chunked in equally sized blocks and rounded up to the next multiple of $K$, where $K$ represents the number of symbols to be coded together and $M$ the resulting parity symbols.
+2) Take $K$, $S$ spaced blocks and $K$ symbols $s$ from each of the selected blocks at offset 0 and encode into $M$ parity symbols, each placed in new $S$ spaced blocks at the same offset and appended to the end of the sequence of original blocks. Repeat this steps at offset $1*sizeof(GF(p))$ and so on.
+3) Repeat the above steps for the length of the entire dataset.
+
+Bellow is a graphical outline of the process:
+
+The sequence of original blocks
+
+![](./figs/blocks.svg)
+
+The logical matrix resulting from stacking each $S$ symbols together, where $K=3$ and $S=4$
+
+![](./figs/matrix1.svg)
+
+Each cell corresponds to a  symbol $s$ and each column of symbols are coded together
+
+![](./figs/matrix3.svg)
+
+The resulting matrix with parity blocks added, where $K=3$, $M=2$, $N=5$ and $S=4$
+
+![](./figs/matrix2.svg)
+
+The resulting structure is a matrix of height $N$ and width $S$. It allows loosing any $M$ "rows" of the matrix and still reconstructing the original dataset. Of course, loosing more than $M$ symbols in a given column would still render the entire dataset invalid, but the likelihood of that happening is mitigated by placing the individual original and parity blocks on independent locations.
+
+Moreover, the resulting dataset is still systematic, which means that the original blocks are unchanged and can be used without prior decoding.
+
+## Data placement and erasures
+
+The code introduced in the above section satisfies our original requirement of any $K$ blocks allowing to reconstruct the original dataset. In fact, one can easily see that every row in the resulting matrix is a standalone element and any $K$ rows will allow reconstructing the original dataset.
+
+However, the code is $K$ rows strong only if we operate under the assumption that each failure is independent of each other. Thus, it is a requirement that each row is placed independently. Moreover, the overall strength of the code decreases based on the number of dependent failures. If we place $N$ rows on two independent locations, we can only tolerate $M=N/2$ failures, three will allow tolerating $M=N/3$ failures and so on. Hence, the code is only as strong as the number of independent locations each element is stored on. Thus, each row and better yet, each element (block) of the matrix should be stored independently and in a pattern mitigating dependence, meaning that placing elements of the same column together should be avoided.
+
+### Load balancing retrieval
```

Comment:
Very clear section. Maybe we can extend it by exploring some placing strategies we discussed with Bulat.

Reply:
I volunteered to write a text about the 3 placing strategies, with ultimate goal to make a post for our blog